### ISSUE TITLE
Post requests with nested objects in body cause no-response

### ISSUE THREAD

**Issue by lomegg** [seriousme/fastify-swaggergen#11]
Hi! Appreciate your work. I am having an issue with your module and I am unable to understand what is happening.

I have a schema defined below. When I use swaggergen to create a route with it, POST requests cause some strange error that results in request never ending.
I spent all day trying to figure this out and I still have no idea.
Handler is there when I check the request. But request never emits end or data.

It drives me crazy.
When I use `fastify.route()` using the same object, it only works fine and returns response. But I cannot use the same url because it has been already taken by non-working route.

So I end up with either undocumented or non-working endpoint.

Please take a look and try to reproduce:
```
fastify.route({
    method: 'POST',
    url: '/test/session/generate',
    schema:
     { tags: [ 'test-session' ],
       summary: 'Generate test session',
       description: 'Generate basic test session in storage',
       operationId: 'TestSession_generate_post',
       produces: [ 'application/json' ],
       body:
        { type: 'object',
          description: 'Incoming Session Schema',
          properties:
           { session_id: { type: 'string', description: 'Session ID' },
             session:
              { type: 'object',
                description: 'Nested Session object',
                properties:
                 { device:
                    { type: 'object',
                      properties:
                       { device_string: { type: 'string' },
                         device_id: { type: 'string' },
                         phone_id: { type: 'string' },
                         uuid: { type: 'string' },
                         advertising_id: { type: 'string' },
                         session_id: { type: 'string' },
                         cookie_string: { type: 'string' } } },
                   cookies:
                    { type: 'object',
                      properties:
                       { mobile:
                          { type: 'object',
                            properties:
                             { mid: { type: 'string' },
                               rur: { type: 'string' },
                               sessionid: { type: 'string' },
                               is_enabled: { type: 'boolean' },
                               user_id: { type: 'string' },
                               csrftoken: { type: 'string' } } },
                         web:
                          { type: 'object',
                            properties:
                             { mid: { type: 'string' },
                               rur: { type: 'string' },
                               sessionid: { type: 'string' },
                               is_enabled: { type: 'boolean' },
                               user_id: { type: 'string' },
                               csrftoken: { type: 'string' } } } } },
                   custom: { type: 'object' } } } } },
       response:
        { '200':
           { type: 'object',
             required: [ 'status' ],
             properties:
              { status:
                 { type: 'string',
                   description: 'Status of request',
                   enum: [ 'ok', 'error', 'fail' ] },
                message: { type: 'string', description: 'Message', example: 'Test' },
                data:
                 { type: 'object',
                   description: 'All data',
                   properties:
                    { wait:
                       { type: 'string',
                         description: 'Need wait seconds for continue',
                         default: '0',
                         example: '20' } } } } } } },
      handler: function (request, reply) {
        reply.send({ hello: 'world', status: 'Ok' })
      }
  });
```

 ```
paths:
  /test/session/generate:
    post:
      security:
        - APIKeyHeader: []
      tags:
        - "test-session"
      operationId: TestSession_generate_post
      summary: "Generate test session"
      description: "Generate basic test session in storage"
      produces:
        - "application/json"
      parameters:
        - in: body
          name: session
          schema:
            $ref: "#/definitions/Session_not_strict"
      responses:
        200:
          $ref: "#/responses/ApiResponse"
        404:
          $ref: "#/responses/NotFoundError"

definitions:
  Device_not_strict:
    type: object
    properties:
      device_string:
        type: string
      device_id:
        type: string
      phone_id:
        type: string
      uuid:
        type: string
      advertising_id:
        type: string
      session_id:
        type: string
      cookie_string:
        type: string

  Cookies_not_strict:
    type: object
    properties:
      mid:
        type: string
      rur:
        type: string
      sessionid:
        type: string
      is_enabled:
        type: boolean
      user_id:
        type: string
      csrftoken:
        type: string

  Session_not_strict:
    type: object
    description: Incoming Session Schema
    properties:
      session_id:
        type: string
        description: Session ID
      session:
        type: object
        description: Nested Session object
        properties:
          device:
            $ref: "#/definitions/Device_not_strict"
          cookies:
            type: object
            properties:
              mobile:
                $ref: "#/definitions/Cookies_not_strict"
              web:
                $ref: "#/definitions/Cookies_not_strict"
          custom:
            type: object
```

**Comment by seriousme** [comment]
Hi,

thanks for the compliment :-)

I tried to validate your swagger/OpenApi schema using the swagger editor at https://editor.swagger.io/ and it complains about the following sections:

```
security:
       - APIKeyHeader: []
```
 but I doubt that would cause your issues.

More important seems to be that some refs are not resolvable:
```
      200:
         $ref: "#/responses/ApiResponse"
       404:
         $ref: "#/responses/NotFoundError"
```
Which could explain why fastify gets stuck on the reply.

If fixing those doesn't work can you please post the full Swagger YAML/JSON  that you use to generate the routes so I can try to reproduce the error ?

Kind regards,
Hans

ps. fastify-swaggergen has been replaced by https://github.com/seriousme/fastify-openapi-glue (see [homepage](https://github.com/seriousme/fastify-swaggergen)) so if possible I would encourage you to switch modules.

**Comment by lomegg** [comment]
Hi! Sorry for delay. Here's the source:
[test.yml.zip](https://github.com/seriousme/fastify-swaggergen/files/3045776/test.yml.zip)

I tried fastify-openapi-glue but the same happens. I suspected error was caused by serializer, but `reply.serializer(JSON.stringify);` does no good. On the controller side, even the instant `reply.send({status: 'ok'})` does no good.

**Comment by seriousme** [comment]
Hi,

Thanks for supplying the spec file.
I did the following:
```
npx github:seriousme/fastify-openapi-glue test.yml
```
This creates a folder called `generatedProject` and in that folder (among other files) a file called `service.js` which contains near the bottom:
```javascript
  async TestSession_generate_post(req, reply) {
    console.log("TestSession_generate_post", req.params);
    return { key: "value" };
  }
```
This is the automatically generated code. However your Swagger definition specifies a different response so I changed the code above to match the Swagger definition (the response from the service needs to match the specification !):
```javascript
  async TestSession_generate_post(req, reply) {
    console.log("TestSession_generate_post", req.params);
    return { 
      "status": "ok",
      "message": "Test",
      "data": {}
    };
  }
```
then I did the following (while being in the `generatedProject folder`:
```
npm install
npm start
```
this installs all required dependencies and runs the webserver.
Then from another terminal I run:

```curl
curl -X POST "http://localhost:3000/test/session/generate" -H "accept: application/json" -H "Content-Type: application/json" -d "{ \"session_id\": \"string\", \"session\": { \"device\": { \"device_string\": \"string\", \"device_id\": \"string\", \"phone_id\": \"string\", \"uuid\": \"string\", \"advertising_id\": \"string\", \"session_id\": \"string\", \"cookie_string\": \"string\" }, \"cookies\": { \"mobile\": \"string\", \"web\": \"string\" }, \"custom\": {} }}"

```

Mind you: I use HTTP and not HTTPS as I haven't configured fastify to use TLS
And then the response I get is:

```
{"status":"ok","message":"Test","data":{}}

```

So it works for me.
Does this solve your problem as well ?

Kind regards,
Hans

**Comment by lomegg** [comment]
I finally figured out what the real reason was. It wasn't the swaggergen.
In my code, I used `swagger-tools` and the first thing to register was `fastify.use(middleware.swaggerMetadata());`
As soon as I commented this out, everything worked like a charm. It seemed to me that I already tried this, but out of all the lines this one was somehow missed.

I didn't figure out yet what exactly caused the issue with POST requests in this plugin, I feel too exhausted to do it. Maybe one day I will, since the issue may come back to bite me in the rear if I don't.

Thank you for your cooperation! Really appreciate.
